File: GLLara/GLLBoneController.m
//
//  GLLBoneController.m
//  GLLara
//
//  Created by Torsten Kammer on 01.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLBoneController.h"

#import "GLLBoneListController.h"
#import "GLLItem.h"
#import "GLLItemBone.h"
#import "GLLModelBone.h"

@interface GLLBoneController ()
{
	id parentController;
}

@property (nonatomic, readonly) NSArray<GLLBoneController *> *childBoneControllers;

@end

@implementation GLLBoneController

- (id)initWithBone:(GLLItemBone *)bone listController:(GLLBoneListController *)listController;
{
	if (!(self = [super init])) return nil;
	
	self.bone = bone;
	self.listController = listController;
	
	return self;
}

- (id)representedObject
{
	return self.bone;
}

- (id)parentController
{
	if (!parentController)
	{
		if (self.bone.parent)
			parentController = [self.listController.boneControllers filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"bone == %@", self.bone.parent]][0];
		else
			parentController = self.listController;
	}
	return parentController;
}

- (NSArray *)childBoneControllers
{
    BOOL hideUnusedBones = [[NSUserDefaults standardUserDefaults] boolForKey:@"hideUnusedBones"];
    if (hideUnusedBones) {
        return [self.listController.boneControllers filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^(GLLBoneController *controller, id dict) {
            GLLItemBone *bone = controller.bone;
            if ([bone.bone.name hasPrefix:@"unused"])
                return NO;
            
            if (bone.parent == self.bone)
                return YES;
            else if (bone.parent == nil)
                return NO;
            
            return [self isAncestorOfUnnamedBone:bone.parent];
        }]];
    } else {
        return [self.listController.boneControllers filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"bone.parent == %@", self.bone]];
    }
}

- (BOOL)isAncestorOfUnnamedBone:(GLLItemBone *)bone {
    if (bone == self.bone)
        return YES;
    else if (![bone.bone.name hasPrefix:@"unused"])
        return false;
    else if (bone.parent)
        return [self isAncestorOfUnnamedBone:bone.parent];
    return NO;
}

#pragma mark - Outline View Data Source

- (id)outlineView:(NSOutlineView *)outlineView child:(NSInteger)index ofItem:(id)item
{
	return self.childBoneControllers[index];
}

- (id)outlineView:(NSOutlineView *)outlineView objectValueForTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
    NSString *string;
    if (self.bone.item != self.listController.item) {
		string = [NSString stringWithFormat:NSLocalizedString(@"%@ (%@)", @"Bone from other model"), self.bone.bone.name, self.bone.item.displayName];
    }
	else
		string = self.bone.bone.name;
    
    // Make bold if bone has non-default values.
    if (self.bone.hasNonDefaultTransform) {
        NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:string attributes:@{NSFontAttributeName : [NSFont boldSystemFontOfSize:[NSFont systemFontSize]]}];
        return attributedString;
    }
    return string;
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(id)item
{
	return self.childBoneControllers.count > 0;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(id)item
{
	return self.childBoneControllers.count;
}

@end

File: GLLara/GLLBoneController.swift
//
//  GLLBoneController.m
//  GLLara
//
//  Created by Torsten Kammer on 01.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Cocoa

/**
 * @abstract Source list controller for a bone.
 * @discussion  Gets its children from the parent bone list controller. If it
 * came from a child item, the name shown is altered to reflect this.
 */
@objc class GLLBoneController: NSObject, NSOutlineViewDataSource {
    let bone: GLLItemBone
    @objc var representedObject: GLLItemBone {
        return bone
    }
    @objc weak var listController: GLLBoneListController!
    
    init(bone: GLLItemBone, listController: GLLBoneListController) {
        self.bone = bone
        self.listController = listController
    }
    
    private weak var parentControllerCached: AnyObject?
    @objc var parentController: AnyObject {
        if let parent = parentControllerCached {
            return parent
        }
        if let parent = bone.parent {
            let parentController = listController.boneControllers.first { ($0 as! GLLBoneController).bone == parent }
            parentControllerCached = parentController as AnyObject?
            return parentController as AnyObject
        } else {
            parentControllerCached = listController
            return listController
        }
    }
    
    @objc var childBoneControllers: [GLLBoneController] {
        let hideUnused = UserDefaults.standard.bool(forKey: "hideUnusedBones")
        if hideUnused {
            return listController.boneControllers.filter({ item in
                guard let controller = item as? GLLBoneController else {
                    return false
                }
                if controller.bone.bone.name.hasPrefix("unused") {
                    return false
                }
                return isDescendantSkippingUnknown(bone: controller.bone)
            }) as! [GLLBoneController]
        } else {
            return listController.boneControllers.filter({ item in
                guard let controller = item as? GLLBoneController else {
                    return false
                }
                return controller.bone.parent == bone
            }) as! [GLLBoneController]
        }
    }
    
    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
        return childBoneControllers[index]
    }
    
    func outlineView(_ outlineView: NSOutlineView, objectValueFor tableColumn: NSTableColumn?, byItem item: Any?) -> Any? {
        let result: String
        if bone.item != listController.item {
            result = String(format: NSLocalizedString("%@ (%@)", comment: "Bone from other model"), bone.bone.name, bone.item.displayName)
        } else {
            result = bone.bone.name
        }
        
        if bone.hasNonDefaultTransform {
            return AttributedString(result, attributes: AttributeContainer([.font : NSFont.boldSystemFont(ofSize: NSFont.systemFontSize)]))
        } else {
            return result
        }
    }
    
    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        return childBoneControllers.count
    }
    
    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
        return !childBoneControllers.isEmpty
    }
    
    func isDescendantSkippingUnknown(bone otherBone: GLLItemBone) -> Bool {
        guard let parent = otherBone.parent else {
            return false
        }
        
        if parent == bone {
            return true
        }
        if otherBone.bone.name.starts(with: "unknown") {
            return isDescendantSkippingUnknown(bone: parent)
        }
        return false
    }
}

